pyggi/line/line.py

```diff
@@ -37,7 +23,9 @@ def __init__(self, target, ingredient):
         self.ingredient = ingredient
 
     def apply(self, program, new_contents, modification_points):
-        return program.do_replace(self, new_contents, modification_points)
+        engine = program.engines[self.target[0]]
+        assert engine == program.engines[self.ingredient[0]]
```

Comment:
Unless the ingredient is modified after the edit creation (and even so), the assert can be moved in the `create()` method so it is checked only once.

Reply:
but the edit instance can be created via not only `create` but also `__init__` method. how about add `is_valid_for(self, program)` methods to the edit classes and check the validity in the method `add` of `Patch`?

Reply:
Unless mistaken as for now the only way PyGGI instantiate new edits is through the `create` method, but you are right the assert is probably better moved directly in `__init__`.
In my opinion checking validity in the `add` method of `Patch` would only make sense if it depended of the other edits.

Reply:
@bloa But there's one problem😞. We cannot check the validity in `__init__` since we don't know the program instance to which the edit will be applied at that moment.

Reply:
You are right again, the program is not known in `__init__`, this is quite unfortunate.
The different possibilities are as follows:
- leave the assert in the `apply` method, which means coherence is checked again and again for no particular reason;
- move the assert in `create`, meaning that edits directly user-instantiated (via `__init__`) are not validated
- adding the program as optional argument to `__init__`, for the only sake of validation
- have a dedicated validation method, called from somewhere else (as you mentioned, presumably `add` in Patch)
- simply remove the assert

Ideally, I would rather have a system that can be proved to not fail than one that checks that it did indeed not fail.
For an edit to be valid, we need first existence of the modification points, validity of the operation for the associated engines, and coherence of the engines if multiple modification points are involved.
Existence can be more or less inferred from the `random_target` method, i don't think validity is neither checked nor inferred, and coherence is the present assert.
(Actually, on top of that, validity inside a patch also depends on the previous edits.)

The best solution I can think of would be to restrict in the `create` method (or in the `random_target` nethod really) the target file to one for which the engine is compatible (and we now have a shared abstract engine class, perfect).
Basically, in simplified pseudo code:
```
def create(cls, program):
    target = program.random_target(AbstractTreeEngine)
    ingredient = program.random_target(program.get_engine(target))
    return cls(target, ingredient)
```
This would ensure that PyGGI will not fail due to incoherent engines.
As for edits directly created using `__init__`, I would simply do nothing: if a user wants to bypass `create` then it is its own responsibility not to provide invalid arguments (in which case PyGGI will fail anyway, assert or not).

Reply:
please check `https://github.com/coinse/pyggi/pull/4/commits/a37dcb55d7616e8f810408d5137ca85067f02493`.